packages/core/src/indexing/client.ts

```diff
@@ -0,0 +1,998 @@
+          if (cachedResult !== undefined) {
+            cache.get(chainId)!.set(getCacheKey(request.request), cachedResult);
+          } else if (request.ev > 0.6) {
+            const resultPromise = requestQueue
+              .request(request.request as EIP1193Parameters<PublicRpcSchema>)
+              .then((result) => JSON.stringify(result))
+              .catch((error) => error as Error);
+
+            // Note: Unawaited request added to cache
+            cache
+              .get(chainId)!
+              .set(getCacheKey(request.request), resultPromise);
+          }
+        }
+
+        if (dbRequests.length > 0) {
+          common.logger.debug({
+            service: "rpc",
+            msg: `Pre-fetched ${dbRequests.length} ${network.name} RPC requests`,
+          });
+        }
+      }
+    },
+    clear() {
+      for (const network of indexingBuild.networks) {
+        cache.get(network.chainId)!.clear();
+      }
+    },
+    set event(_event: Event | SetupEvent | undefined) {
+      event = _event;
+    },
+  };
+};
+
+export const cachedTransport =
+  ({
+    common,
+    network,
+    requestQueue,
+    syncStore,
+    cache,
+  }: {
+    common: Common;
+    network: Network;
+    requestQueue: RequestQueue;
+    syncStore: SyncStore;
+    cache: Cache;
+  }): Transport =>
+  ({ chain }) =>
+    custom({
+      async request({ method, params }) {
+        const body = { method, params };
+
+        // multicall
+        if (
+          method === "eth_call" &&
+          params[0]?.data?.startsWith(MULTICALL_SELECTOR)
+        ) {
+          let blockNumber: Hex | "latest" | undefined = undefined;
+          [, blockNumber] = params;
+
+          const multicallRequests = decodeFunctionData({
+            abi: multicall3Abi,
+            data: params[0]!.data,
+          }).args[0];
+
+          if (multicallRequests.length === 0) {
+            // empty multicall result
+            return "0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000";
+          }
+
+          const requests = multicallRequests.map(
+            (call) =>
+              ({
+                method: "eth_call",
+                params: [
+                  {
+                    to: call.target,
+                    data: call.callData,
+                  },
+                  blockNumber,
+                ],
+              }) as const satisfies EIP1193Parameters,
+          );
+          const results = new Map<
+            EIP1193Parameters,
+            {
+              success: boolean;
+              returnData: `0x${string}`;
+            }
+          >();
+          const requestsToInsert = new Set<EIP1193Parameters>();
+
+          for (const request of requests) {
+            const cacheKey = getCacheKey(request);
+
+            if (cache.get(network.chainId)!.has(cacheKey)) {
+              const cachedResult = cache.get(network.chainId)!.get(cacheKey)!;
```

Comment:
Is it possible for something to be in the cache that was _not_ a prediction for the current batch of events?

Reply:
no